File: Sunrin-ICPC-2022/Final-I/FI.py
n = int(input())
a = []
for i in range(n): a.append(list(map(int, input().split())))
d = [[0 for c in range(n)] for r in range(n)]
res = 0

for i in range(n):
    for j in range(n):
        mx = 0
        for r in range(i+1):
            for c in range(j+1):
                if a[r][c] < a[i][j]: mx = max(mx, d[r][c])
        d[i][j] = mx + 1
        res = max(res, d[i][j])
print(res)

File: Sunrin-ICPC-2022/Final-I/FI.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().split('\n');

const n = +input[0];
let a = [];
for(let i=1; i<=n; i++) a.push(new Array(n));
for(let i=1; i<=n; i++){
    const tmp = input[i].split(' ');
    for(let j=0; j<n; j++) a[i-1][j] = +tmp[j];
}
let d = [], res = 0;
for(let i=0; i<n; i++) d.push(new Array(n));
for(let i=0; i<n; i++){
    for(let j=0; j<n; j++){
        let mx = 0; d[i][j] = 0;
        for(let r=0; r<=i; r++) for(let c=0; c<=j; c++) if(a[r][c] < a[i][j]) mx = Math.max(mx, d[r][c]);
        d[i][j] = mx + 1;
        res = Math.max(res, d[i][j]);
    }
}
console.log(res);